Show progress on kernel building

agustinhenze opened the pull request:
Try to follow PR #30 but with a different approach plus test cases.

stefwalter commented:
From experience with reliable automated infrastructure, having a 60 minute timeout is risky. Could you bump it to 10 or 12 hours? We really want to catch the pathological case here.

spbnick commented on skt/__init__.py:
```diff
@@ -503,24 +508,53 @@ def mktgz(self, clean=True):
+            if not timer.isAlive():
+                raise CommandTimeoutError(
+                    "'%s' was taking too long" % ' '.join(args)
+                )
```
It seems to me this can race with the timer handler (`make.terminate`) and produce an error even if `make` has finished on time. For example, if it finished during that `time.sleep(1)` or during blocked output to stdout. Could you perhaps write a timer handle which runs `make.poll()` and sets a variable if `make` was still alive on timeout, and *then* calls `make.terminate()`? Then you can check that variable and throw an error if it's set.

BTW, shouldn't we also be checking make's exit status?

agustinhenze replied:
Yes, it wasn't checking the returncode. I have added some test cases that should validate the correct behavior. Please take a look at it again.

spbnick replied:
Thanks a lot :smiley:! However, timer could still have expired right after make failed, but before we checked, and then we would report the wrong error. Do you think you can fix that?

agustinhenze replied:
oh, I didn't get it before. Good catch! Please take a look at it now.

spbnick commented on skt/__init__.py:
```diff
@@ -498,6 +498,17 @@ def getrelease(self):
+        Returns:
+            The full path of the tarball generated.
+        Raises:
+            CommandTimeoutError: When building kernel takes too long.
```
I think this needs to say "takes longer than the specified timeout".

agustinhenze replied:
Done

spbnick commented on skt/__init__.py:
```diff
@@ -503,24 +508,53 @@ def mktgz(self, clean=True):
+
+            timer = Timer(timeout, make.terminate)
+            # If the process is cancelled, it doesn't need to wait for anything
+            timer.setDaemon(True)
```
Thanks for the comment above, but what does this line actually do? I'm unfamiliar with this API, and wasn't able to find documentation on this quickly.

agustinhenze replied:
I have rewritten the comment, I hope it's understandable now

spbnick replied:
Sorry, it's probably obvious to regular Python developers (and if that's true, perhaps no comment is necessary), but not to me. What's the difference between `setDaemon(False)` and `setDaemon(True)` in timer behavior? A link to documentation would also be fine. Thanks :)

agustinhenze replied:
The difference in terms of behavior between a daemon timer and a non-daemon timer is none. Timer is a subclass of Thread, that means it is necessary to read the thread section https://docs.python.org/2/library/threading.html#thread-objects, I have to say that the doc is not clear here... The long story short for our case is, if we don't set as daemon this timer, the main process cannot be stopped (e.g. Ctrl+C or SIGTERM) because it will keep waiting until the timer/thread finish. We can solve this too, calling timer.cancel() on SIGTERM or Ctrl+C but I think this way is much simpler.

Now that you mention it, I think it'd be better remove the comment here...

spbnick replied:
I see, thank you for the explanation :)

agustinhenze commented:
> From experience with reliable automated infrastructure, having a 60 minute timeout is risky. Could you bump it to 10 or 12 hours? We really want to catch the pathological case here.

@stefwalter ok, bumped to 12 hours. Please take a look at it and it'd be great if you test it a little bit as well.

agustinhenze commented:
> Also, we can have documentation addition in the same commit, as this technically changes the function interface, and documentation should be updated together with that.

Squashed

agustinhenze commented on skt/__init__.py:
```diff
@@ -503,24 +508,53 @@ def mktgz(self, clean=True):
+
+class CommandTimeoutError(Exception):
+    """
+    This exception is raised when a timeout occurs on a process which has had
```
Done

spbnick commented on skt/__init__.py:
```diff
@@ -503,24 +524,68 @@ def mktgz(self, clean=True):
+class RegexNotFound(Exception):
+    """
+    Exception raised when a regex does not match and it is impossible to
+    continue. The accompanying value is a string whose explains what it can not
```
I think this should say "which", not "whose".

spbnick commented on skt/__init__.py:
```diff
@@ -503,24 +524,68 @@ def mktgz(self, clean=True):
-        if tgzpath is not None:
-            fpath = "/".join([self.path, tgzpath])
+        if not os.path.isfile(fpath):
+            raise IOError("No such file {}".format(fpath))
```
Can you elaborate here a little? E.g. say "Built kernel tarball {} not found"?

spbnick commented on skt/__init__.py:
```diff
@@ -493,8 +497,25 @@ def getrelease(self):
+        Raises:
+            CommandTimeoutError: When building kernel takes longer than the
+                                 specified timeout.
+            CalledProcessError:  When a command returns an exit code different
```
Shouldn't this say "When "make" returns an exit code..."? Or do you want to keep that abstracted away, or plan adding other commands to this?

agustinhenze replied:
Yes, I'd like to keep it abstracted because it calls to `.prepare` which uses check_call and the latter raises this exception is something goes wrong.

agustinhenze commented:
I think, actually I'm pretty sure, that this commit 0d3c31f626a0bcbc1f2d1dfd759c1669d5dfb2ff broke master.
```
2018-03-28 19:18:12,803     INFO   basecfg: tests/smoke/kernel-minimal-config
2018-03-28 19:18:12,803     INFO   cfgtype: olddefconfig
2018-03-28 19:18:12,803    DEBUG   state: buildlog -> ../skt-workdir/build.log
Traceback (most recent call last):
  File "skt/executable.py", line 627, in <module>
    main()
  File "skt/executable.py", line 616, in main
    args.func(cfg)
  File "skt/executable.py", line 105, in wrapper
    func(cfg)
  File "skt/executable.py", line 184, in cmd_build
    raise e
AttributeError: 'kbuilder' object has no attribute 'enable_debuginfo'
```

veruu commented:
> I think, actually I'm pretty sure, that this commit 0d3c31f broke master.

Thanks for reporting, I submitted a pull request to add the missing attribute.

spbnick commented on skt/__init__.py:
```diff
@@ -518,24 +539,73 @@ def mktgz(self, clean=True):
+            make = subprocess.Popen(args,
+                                    stdout=writer,
+                                    stderr=subprocess.STDOUT)
+            make.timedout = False
```
Hmm, is it normal to just go and add a field to an instance of some 3rd-party class? What if a later version of subprocess adds a field named the same, but used for something else?

agustinhenze replied:
You're right! It is not a good idea. Anyway, I was looking at the code again and... The race still there, because if the process ends after the poll call, it will report a timeout error when the process finished correctly... So I think I'm going back to the first implementation which it is simpler. What do you think?

spbnick replied:
Well, if the process was still running at the moment of the poll call, it means it timed out. I.e. it took longer than allowed. So, I think that part is OK.

agustinhenze replied:
Fair enough, please take a look at it again.

spbnick replied:
OTOH, previously you simply checked if the timer has expired or not, and that might have happened after make has finished. E.g. because of the sleep in the loop, or because your output to stdout was blocked.

spbnick commented on skt/__init__.py:
```diff
@@ -518,24 +539,73 @@ def mktgz(self, clean=True):
+    Exception raised when a regex does not match and it is impossible to
+    continue. The accompanying value is a string which explains what it can not
+    find.
+    """
```
Do you think you can name this something like `ParsingError`, or similar? The `RegexNotFound` is a bit too specific, while at the same time not that helpful.

agustinhenze replied:
OK, I'll rename it to `ParsingError` :+1:

spbnick commented on skt/__init__.py:
```diff
@@ -518,24 +539,73 @@ def mktgz(self, clean=True):
+                self.append_and_log2stdout(reader.readlines(), stdout_list)
+                time.sleep(1)
+            self.append_and_log2stdout(reader.readlines(), stdout_list)
+            timer.cancel()
```
Could you wrap the code executing while the timer is running, above, into a try-finally, and put `timer.cancel` into the `finally`, so we're sure the timer gets killed?

agustinhenze replied:
Great idea! :+1:

spbnick commented on tests/test_init.py:
```diff
@@ -53,3 +60,99 @@ def test_parse_bad_patchwork_url(self):
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir)
+
+    def test_clean_mktgz(self):
```
Could you name those so that they all start with `test_mktgz_` and then have specifics on the end? E.g. `test_mktgz_clean`?

spbnick commented on tests/test_init.py:
```diff
@@ -53,3 +60,99 @@ def test_parse_bad_patchwork_url(self):
+        shutil.rmtree(self.tmpdir)
+
+    def test_clean_mktgz(self):
+        """Check if `mrproper` target is called when clean is True"""
```
This checks much more than just the `mrproper` target being "made". Please remove the rest of the tests from here, and please also only check that there is a command which has `make` as the first argument and `mrproper` as one of the following arguments. Otherwise this test is going to break often (including when you rebase now).

agustinhenze replied:
Ok, done

spbnick commented on tests/test_init.py:
```diff
@@ -53,3 +60,99 @@ def test_parse_bad_patchwork_url(self):
+    def test_mktgz(self):
+        """
+        Check if ParsingError, IOError, and subprocess.CalledProcessError are
+        raised, as well also check if mktgz can finish successfully.
```
Do you think you can split this into separate test functions?

agustinhenze commented:
> Thank you, Agustin, this is a needed change, and sorry for the delay with the review. Please take a look at my comments, and also please rebase on the latest master. Thank you!

@spbnick Please take a look at it again. I have addressed all your comments and rebased to master.

spbnick commented on tests/test_init.py:
```diff
@@ -53,3 +60,102 @@ def test_parse_bad_patchwork_url(self):
+            self.assertNotEqual(
+                m_check_call.mock_calls[0],
+                mock.call(['make', '-C', self.tmpdir, 'mrproper']),
+            )
```
This still tests the exact arguments, which is making the test fragile. Do you think you can test that argv[0] == "make" and argv[something above zero] == "mrproper", and the reverse?

spbnick commented on tests/test_init.py:
```diff
@@ -53,3 +60,102 @@ def test_parse_bad_patchwork_url(self):
+        self.m_io_open.__enter__ = lambda *args: self.m_io_open
+        self.m_io_open.__exit__ = lambda *args: None
+        self.ctx_io_open = mock.patch('io.open',
+                                      Mock(return_value=self.m_io_open))
```
I'm concerned about the approach of having instance variables shared and updated by tests. This is similar to having global variables. I think this creates an opportunity for a test to use one of those without updating it, and then breaking when tests are reordered or omitted. Do you think you can instead, e.g., write functions creating those values as needed and then use them in the tests, or just directly creating them every time?

spbnick commented on tests/test_init.py:
```diff
@@ -53,3 +60,99 @@ def test_parse_bad_patchwork_url(self):
+            self.assertIn('mrproper', m_check_call.mock_calls[0][1][0])
+            m_check_call.reset_mock()
+            self.assertRaises(Exception, self.kbuilder.mktgz, clean=False)
+            self.assertNotIn('make', m_check_call.mock_calls[0][1][0])
```
Strictly-speaking, this should check that `make` is not in the first argument only, not in any position. And the reverse above.

agustinhenze replied:
I think it's **not** going to be better than the previous version in terms of the test fragility, after all, it's checking the exact arguments using hardcoded indexing. And the output when it fails is less intuitive. So I propose two things here, or we remove this test case or back to the previous version.

spbnick replied:
Hmm, perhaps I misunderstood what's being done here, I'm not familiar with Python's `Mock`. I read this specific line (100) as it's checking that there was a command executed, which had no exact string `make` as any element in the `argv` array. E.g. command "echo hello" would pass, command "make something" would fail, but command "echo make" would fail as well and the latter was my concern.

However, I see that this is not as trivial as I thought. Could you perhaps explain what you're doing here, and why you think changing it to what I suggest would be worse? Thank you.

agustinhenze replied:
Yeah, np. In the beginning, you mentioned that the test case was fragile because it checked all the arguments and I got your point. But if you take a look of what we'll have now (after addressing your last comment) and what we had before, it doesn't seem to be better at least for me in the term of fragility.

I didn't push the following code yet. It retrieves the first call to subprocess.check_call with the anti-intuitive [0][1][0] indexing and then compares index [0] == 'make' and so on.
```python
            self.assertEqual('make', m_check_call.mock_calls[0][1][0][0])
            self.assertIn('mrproper', m_check_call.mock_calls[0][1][0][3])
..
            self.assertNotEqual(make', m_check_call.mock_calls[0][1][0][0])
            self.assertNotEqual('mrproper', m_check_call.mock_calls[0][1][0][3])
```
vs
```python
            self.assertEqual(
                m_check_call.mock_calls[0],
                mock.call(['make', '-C', self.tmpdir, 'mrproper'])
            )
...
            self.assertNotEqual(
                m_check_call.mock_calls[0],
                mock.call(['make', '-C', self.tmpdir, 'mrproper']),
            )
```
Both have hardcoded arguments and for me the previous one has a better output when it fails.
```
======================================================================
FAIL: test_mktgz_clean_2 (tests.test_init.KBuilderTest)
Check if `mrproper` target is called when clean is True
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_init.py", line 109, in test_mktgz_clean_2
    mock.call(['mke', '-C', self.tmpdir, 'mrproper'])
AssertionError: call(['make', '-C', '/tmp/tmpQ1LiQC', 'mrproper']) != call(['mke', '-C', '/tmp/tmpQ1LiQC', 'mrproper'])
```
vs
```
======================================================================
FAIL: test_mktgz_clean (tests.test_init.KBuilderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_init.py", line 96, in test_mktgz_clean
    self.assertEqual('mke', m_check_call.mock_calls[0][1][0][0])
AssertionError: 'mke' != 'make'
```
I don't think it is worse, but I am not convinced it is better either hehe. I'm used to seeing comparisons using mock.call, maybe that's why it sounds more intuitive to me. Anyway I want to see this PR merged soon so I'll take the way you prefer and if in the future we find a better way to solve this, we can change it :):

spbnick replied:
Ah, I see, thank you, Agustin. OK, minor correction to what I actually
suggested about `mrproper`: it should be searched for among all arguments. So,
it should rather be:

    self.assertEqual('make', m_check_call.mock_calls[0][1][0][0])
    self.assertIn('mrproper', m_check_call.mock_calls[0][1][0])

However, I agree that readability of both assert and the failure output is bad
in this case, so I think it's OK for us to match the exact command line, and
change this test from time to time as we update it, if that means we can
actually understand the test code and its failures.

Take your pick on how you want this done :)

agustinhenze replied:
Thank you Nikolai! I really like your reviews :smile:. Please take a look at it again to see if it is what you expected.

spbnick commented on tests/test_init.py:
```diff
@@ -53,3 +60,103 @@ def test_parse_bad_patchwork_url(self):
+        """
+        self.m_io_open.readlines = Mock(
+            return_value=['foo\n', self.success_str, 'bar']
+        )
```
Hmm, we shouldn't get to actually reading the "make" output if it fails to spawn, therefore this is unnecessary and is confusing. Can we remove this, and the `ctx_io_open` use? Also, `ctx_check_call` shouldn't be used either, right?

agustinhenze replied:
Yeah! Done!

agustinhenze commented:
It seems that pypi is broken.
![screenshot_20180416_124903](https://user-images.githubusercontent.com/1900634/38820527-8c991002-4174-11e8-8b98-5ed2a7499607.png)
I don't understand why travis doesn't stop there and keep trying next commands despite pip install dropped an exit code != 0.

spbnick commented:
@agustinhenze, yeah, something is amiss here. Regarding Travis CI continuing, I think we should put `pip install .` into [`before_script`](https://docs.travis-ci.com/user/customizing-the-build/#Breaking-the-Build). Can you check and submit a PR if that helps?

spbnick commented:
Or perhaps we should use `install` instead: https://docs.travis-ci.com/user/customizing-the-build/#The-Build-Lifecycle

agustinhenze commented:
> Or perhaps we should use install instead: https://docs.travis-ci.com/user/customizing-the-build/#The-Build-Lifecycle

Yeah! That worked! Well pypi still broken but our CI is improved! :smile:

spbnick commented:
Thank you, now we have proper dependency checking, and pip is working again, great :) I tried testing your tests, and there's one thing, which I think needs fixing: the tests produce output on stdout when running, which is not at all helpful and breaks normal unittest output:

    .foo
    Tarball successfully created in ./linux-4.16.0.tar.gz
    ..foo
    bar
    .foo
    Tarball successfully created in ./linux-4.16.0.tar.gz
    bar..........
    ----------------------------------------------------------------------
    Ran 14 tests in 0.044s

    OK

Do you think you can make the tests silent?

agustinhenze commented:
> Do you think you can make the tests silent?

Done

spbnick commented:
Merged, thank you, Agustin!
